src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/OnTypeRenameHandler.cs

```diff
@@ -0,0 +1,111 @@
+            LSPDocumentMappingProvider documentMappingProvider)
+        {
+            if (documentManager is null)
+            {
+                throw new ArgumentNullException(nameof(documentManager));
+            }
+
+            if (requestInvoker is null)
+            {
+                throw new ArgumentNullException(nameof(requestInvoker));
+            }
+
+            if (projectionProvider is null)
+            {
+                throw new ArgumentNullException(nameof(projectionProvider));
+            }
+
+            if (documentMappingProvider is null)
+            {
+                throw new ArgumentNullException(nameof(documentMappingProvider));
+            }
+
+            _documentManager = documentManager;
+            _requestInvoker = requestInvoker;
+            _projectionProvider = projectionProvider;
+            _documentMappingProvider = documentMappingProvider;
+        }
+
+        public async Task<DocumentOnTypeRenameResponseItem> HandleRequestAsync(DocumentOnTypeRenameParams request, ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (!_documentManager.TryGetDocument(request.TextDocument.Uri, out var documentSnapshot))
+            {
+                return null;
+            }
+
+            var projectionResult = await _projectionProvider.GetProjectionAsync(documentSnapshot, request.Position, cancellationToken).ConfigureAwait(false);
+            if (projectionResult is null || projectionResult.LanguageKind != RazorLanguageKind.Html)
+            {
+                return null;
+            }
+
+            var onTypeRenameParams = new DocumentOnTypeRenameParams()
+            {
+                Position = projectionResult.Position,
+                TextDocument = new TextDocumentIdentifier() { Uri = projectionResult.Uri }
+            };
+
+            var contentType = projectionResult.LanguageKind.ToContainedLanguageContentType();
+            var response = await _requestInvoker.ReinvokeRequestOnServerAsync<DocumentOnTypeRenameParams, DocumentOnTypeRenameResponseItem>(
+                MSLSPMethods.OnTypeRenameName,
+                contentType,
+                onTypeRenameParams,
+                cancellationToken).ConfigureAwait(false);
+
+            if (response is null)
+            {
+                return null;
+            }
+
+            var mappingResult = await _documentMappingProvider.MapToDocumentRangesAsync(
+                RazorLanguageKind.Html,
+                request.TextDocument.Uri,
+                response.Ranges,
+                cancellationToken).ConfigureAwait(false);
+
+            if (mappingResult is null ||
+                (_documentManager.TryGetDocument(request.TextDocument.Uri, out var mappedDocumentSnapshot) &&
```

Comment:
>(_documentManager.TryGetDocument(request.TextDocument.Uri, out var mappedDocumentSnapshot)  [](start = 16, length = 91)

This needs to be done again?

Reply:
Yup because it's an immutable document snapshot. It could have changed between the initial retrieval and the async work.

Makes me think we should probably change the document manager's "TryGetDocument" method to be more explicit that you're getting the snapshotted view of the world.